base/array.jl

```diff
@@ -187,12 +185,15 @@ function vect(X...)
     return T[X...]
 end
 
-size(a::Array, d::Integer) = arraysize(a, d isa Int ? d : convert(Int, d))
-size(a::Vector) = (arraysize(a,1),)
-size(a::Matrix) = (arraysize(a,1), arraysize(a,2))
-size(a::Array{<:Any,N}) where {N} = (@inline; ntuple(M -> size(a, M), Val(N))::Dims)
+size(a::Array, d::Integer) = size(a, Int(d)::Int)
+function size(a::Array, d::Int)
+    d < 1 && error("arraysize: dimension out of range")
```

Comment:
This probably shouldn't refer to `arraysize` anymore at least, but might as well make this a more useful error while we're at it?
```suggestion
    d < 1 && throw(ArgumentError(lazy"size: invalid dimension $d for $(ndims(a))d Array")
```
Or are we worried about introducing more compiler-side complexity in this hot path?